src/shims/env.rs

```diff
@@ -44,7 +44,7 @@ fn alloc_env_var<'mir, 'tcx>(
     bytes.push(b'=');
     bytes.extend_from_slice(value);
     bytes.push(0);
-    memory.allocate_static_bytes(bytes.as_slice(), MiriMemoryKind::Env.into())
+    memory.allocate_static_bytes(bytes.as_slice(), MiriMemoryKind::Env.into()).assert_ptr()
```

Comment:
IMO something very weird is going on when an allocation function doesn't return a `Pointer`. We are embedding things like "zero-sized allocations just cast the alignment to a ptr" deep into the Abstract Machine semantics. That's not very elegant at all, my goal was for `exchange_malloc` to be the *only* place that does that.

Reply:
That will only work for heap allocations. I don't see how to handle stack allocations without encoding this into the abstract machine.

Reply:
I don't see why we would make any promises about stack allocations at all.

Reply:
(Also I didn't yet have time to look at your latest changes in https://github.com/rust-lang/rust/pull/67192... but sure when I'll get around to that.)

Reply:
Please wait until I have adjusted that PR to not do unnecessary changes. I think we can drop the entire PR except for the cleanups and make a tiny targetted change to promotion to not change its behaviour instead of messing with all of mir interpret

Reply:
Okay. Maybe 'll even understand the new version. ;)

Notice that we have two existing uses of `MPlaceTy::dangling`, but for both of them the address is not actually observable by the program. So that's less sketchy than using it for allocations the user can see. I'd prefer to get rid of them but I am not sure how...
Probably it would be good to add some comments explaining all this alongside `MPlaceTy::dangling` and the places where it gets used.